imx_pwm_led.c

```diff
@@ -0,0 +1,2069 @@
+	context.mda = self->mmio_base_phys + PWMSAR;
+	/* destination address frozen;
+	 * source address postincrement;
+	 * start in read mode */
+	context.ms = 0x00100000;
+	context.pda = epit_status_register_address(glb_data.epit);
+	/* destination address frozen;
+	 * 32-bit write size; start in write mode */
+	context.ps = 0x000c0400;
+
+	/* acquire the channel and load its context;
+	 * it's triggered externally by the EPIT */
+	sdma_setup_channel(self->sdmac, true);
+
+	ret = sdma_load_partial_context(self->sdmac, &context, 0,
+					sizeof(context));
+	if (ret) {
+		dev_err(self->dev, "%s: failed to load context", __func__);
+		return ret;
+	}
+	return 0;
+}
+
+static int load_sdma_duty(struct imx_pwm_led *self, uint cue_idx, uint led_idx,
+			  uint duty)
+{
+	int ret;
+	struct sdma_context_data context = { { 0 } };
+	uint duty_offset = (cue_idx * max_leds) + led_idx;
+	dma_addr_t duty_phys =
+		glb_data.duty_buf_phys + (duty_offset * sizeof(u16));
+	dev_dbg(self->dev, "%s: led: %u, duty: %u", __func__, led_idx, duty);
+	glb_data.duty_buf[duty_offset] = duty;
+
+	/* Write the ARG_PRELOAD, ARG_END and MSA; as well as the MDA,
+	 * since it's in the middle */
+	context.gReg[ARG_PRELOAD] = 1;
+	context.gReg[ARG_END] = duty_phys + sizeof(u16);
+	context.mda = self->mmio_base_phys + PWMSAR;
+	context.msa = duty_phys;
+
+	/* This operation is run by channel 0, which is mutually exclusive with
+	 * our scripts's channel. Since only one channel is running at a time,
+	 * and channels cannot preempt each other, this ensures that the context
+	 * is only updated between iterations of our script. (i.e. registers
+	 * won't get suddenly updated while our script is running) */
+	ret = sdma_load_partial_context(
+		self->sdmac,
+		(struct sdma_context_data *)&context.gReg[ARG_PRELOAD],
+		offsetof(struct sdma_context_data, gReg[ARG_PRELOAD]),
+		16); /* 4 * sizeof(u32) = 12 bytes */
+	if (ret) {
+		dev_err(self->dev, "%s: failed to load context", __func__);
+		return ret;
+	}
+	return 0;
+}
+
+static int fetch_sdma_duty(struct imx_pwm_led *self, uint *duty)
+{
+	int ret = sdma_fetch_partial_context(self->sdmac, duty,
+					     offsetof(struct sdma_context_data,
+						      gReg[ARG_DUTY]),
+					     4); /* sizeof(u32) = 4 bytes */
+	if (ret) {
+		dev_err(self->dev, "%s: failed to fetch context", __func__);
+		return ret;
+	}
+	dev_dbg(self->dev, "%s: %u", __func__, *duty);
+	return 0;
+}
+
+static int load_sdma_fade(struct imx_pwm_led *self, uint fade_idx)
+{
+	int ret;
+	struct led_fade *fade;
+	struct sdma_context_data context = { { 0 } };
+	u32 offset = 0;
+	dev_dbg(self->dev, "%s: led: %u, fade: %u", __func__, self->led_idx,
+		fade_idx);
+	if ((ret = open_fade(self, fade_idx))) {
+		return ret;
+	}
+	fade = &glb_data.fades[fade_idx];
+	mutex_lock(&fade->lock);
+	if (fade->len == 0) {
+		dev_warn(self->dev, "%s: fade %u: len %u < %u", __func__,
+			 fade_idx, fade->len, sizeof(u16));
+		ret = -ENODATA;
+		goto out;
```

Comment:
No it should be left open for several reasons:
- if `PERM_MODE_PID_OWNER` is used and there are multiple processes handling different LED devices, then it's up to the system designer to assign which fades/cues are being used by which process. Hence there's no need to close the fade/cue, since nobody else should be using it.
- when the device is closed all fades and cues that were opened by the calling process are automatically closed.